### PR TITLE
better supress warnings and errors preprocessor

### PR DESCRIPTION
**Pull request by friendlyhj**

### CODE REVIEW

**Comment by jaredlll08** [comment]
Looks good to me, just want @kindlich thoughts on it

**Comment by kindlich** [comment]
Wasn't nowarn also supposed to suppress errors?
iSome mods printed errors if nonexistant recipes were removed which caused many entries printed to the player log when using a batch method to remove "all" recipes for items.

If I remember correctly, NuclearCraft was one of these mods that printed an error.
Therefore, changing this behavior would cause some packs to suddenly print a lot of errors to the players even though "everything's fine".

**Comment by friendlyhj** [comment]
But it also suppresses very many fatal errors (wrong method argument and so on), which make `/ct syntax` show no error, unless you see logs.

I has doubted for a long time why I cannot be sent error message in the game.

The description of `#ikwid` is not confirm with its actual act. It is NoWarn, not NoWarnAndError.

I think errors should be avoided, not be suppressed.

**Comment by kindlich** [comment]
Well,I would suppose that if you know what you're doing (what `ikwid` stands for) that you'd know to check your log.

While I agree that errors should be avoided, think of something like
```zs
for item in itemUtils.getByRegexReistryName("nc:.*") {
    mods.nuclearcraft.Machine.removeRecipe(item);
    recipes.remove(item);
    mods.othermod.Machine.remove(item);
}
```

That is a proper syntax and proper semantics.
Still it would generate a lot of errors in the chat.
It is far more readable to say `remove all recipes from NC` rather than `remove item A, B, C from machine X, remove crafting recipes for A, F H, remoce otherMachine recipes for A, C, F, H` which is why some packs do that.


The `ikwid` preprocessor is only supposed to be used when you are already done with development and want to suppress false positives from the user.

**Comment by jaredlll08** [comment]
So after thinking about it more, I don't want to break compatibility with previous scripts, so ideally this should stay as it currently is.

I'm not opposed to a `#noWarn` (just change `#ikwid`'s variable name) that just kills warnings but leaves errors.

File: CraftTweaker2-API/src/main/java/crafttweaker/CraftTweakerAPI.java [hunk 1]
```diff
@@ -91,11 +91,14 @@
...
+
     /**
-     * If true warnings sent to the log won't print to the player's chat
+     * 3 bits flag.
+     * if the highest bit is 1, errors and warnings must print to player's chat, ignoring the two other bits.
+     * if the middle bit is 1, errors won't print to the player's chat.
+     * if the lowest bit is 1, warnings won't.
      */
-    public static boolean noWarn = false;
+    public static byte suppressWarnAndErrorFlag = 0b000;
```
Comment (kindlich) [inline CraftTweaker2-API/src/main/java/crafttweaker/CraftTweakerAPI.java]: While using a flag is an interesting approach, how about we go for an enum instead?
Something like
```java
public enum LogLevel {
    NORMAL, NO_WARNING, NO_ERROR
}
```

Edit: Ignore, already resolved, was on an older state

Comment (jaredlll08) [inline CraftTweaker2-API/src/main/java/crafttweaker/CraftTweakerAPI.java]: I would honestly prefer to have an enum for this instead of being bit based.

**Review by jaredlll08 — CHANGES_REQUESTED** [review]

File: CraftTweaker2-API/src/main/java/crafttweaker/api/logger/MTLogger.java [hunk 2]
```diff
@@ -61,14 +57,12 @@ public void logWarning(String message) {
             logger.logWarning(message);
         }
         
-        String message2 = "WARNING: " + message;
+        String message2 = "\u00a7eWARNING: " + message;
```
Comment (kindlich) [inline CraftTweaker2-API/src/main/java/crafttweaker/api/logger/MTLogger.java:60]: That unicode was for coloring the message iirc.
Any reason to remove it?

File: CraftTweaker2-API/src/main/java/crafttweaker/api/logger/MTLogger.java [hunk 3]
```diff
@@ -84,21 +78,25 @@ public void logError(String message, Throwable exception) {
             logger.logError(message, exception);
         }
         
-        String message2 = "ERROR: " + message;
+        String message2 = "\u00a7cERROR: " + message;
```
Comment (kindlich) [inline CraftTweaker2-API/src/main/java/crafttweaker/api/logger/MTLogger.java:81]: That unicode was for coloring the message iirc.
Any reason to remove it?

File: CraftTweaker2-API/src/main/java/crafttweaker/api/logger/MTLogger.java (hunk 2 above)
Comment (friendlyhj) [reply CraftTweaker2-API/src/main/java/crafttweaker/api/logger/MTLogger.java:60]: Coloring warnings and errors message can make them catch the eye. If you want, I can remove `§`.

Comment (kindlich) [reply CraftTweaker2-API/src/main/java/crafttweaker/api/logger/MTLogger.java:60]: Oh, wait, I read that the other way round lmao

**Review by kindlich — APPROVED** [review]
Looks good to me

File: CraftTweaker2-API/src/main/java/crafttweaker/CraftTweakerAPI.java [hunk 4]
```diff
@@ -217,6 +219,30 @@ public static void logError(String message, Throwable exception) {
     public static void logDefault(String message) {
         getLogger().logDefault(message);
     }
+
+    /**
+     * Set whether errors and warnings printed to players' log or not.
```
Comment (jaredlll08) [inline CraftTweaker2-API/src/main/java/crafttweaker/CraftTweakerAPI.java]: When you say `players' log` do you mean their chat, or their actual log file?

If you mean the log file then:
```
Set whether errors and warnings are printed to the players' log or not.
```

If you mean their chat, then:
```
Set whether errors and warnings are printed to the players' chat or not.
```

**Review by jaredlll08 — CHANGES_REQUESTED** [review]

**Review by jaredlll08 — APPROVED** [review]

**Pull request merged**